### PR TITLE
Server shim via tokenization and string manipulation

### PR DESCRIPTION
**Pull request by ggoodman**
## Description

This PR changes the `@twind/shim/server` implementation to use an approach based on handling an html token stream and string manipulation.

The motivation for this PR is that I was observing that the existing `shim()` implementation was resulting in hard-to-understand changes to the input markup. A review of the code showed that it was performing a full parse, manipulating the AST and then reserializing in a potentially lossy way.

This PR uses the `Tokenizer` from [`htmlparser2`](https://github.com/fb55/htmlparser2/blob/master/src/Tokenizer.ts) to identify blocks of the input markup that need to be modified. Since the tokenizer will be scanning through the code and exposes offset information, we can decompose the original markup into alternating 'chunks' of text before an edit and text after the edit. Once the entire markup is scanned, we can reconstruct the modified document from the chunks.

The design of this PR is also motivated by the desire to do less work:

1. If no substitutions are required, no string manipulation (or AST serialization) will be done. People not using grouping will benefit here.
2. Even if substitutions are required, _we do not touch the original markup_ except for the attribute value string itself. When we do make modifications we replace an AST serialization with an `Array#join('')` and a string concatenation + slice.

Due to the removal of `node-html-parser`, some tests related to its nuances were removed.

Also, the options `decodeEntities` and `xmlMode` of the `htmlparser` `Tokenizer` constructor are intentionally _not_ exposed but this could easily be changed. These two options are currently both hard-coded to `false`.

Initial local benchmarking of 1000 warm iterations of the old and new approaches suggests that this PR introduces a ~40% average performance improvement.

### CODE REVIEW

File: src/shim/server/shim.test.ts [hunk 1]
```diff
@@ -134,102 +134,4 @@ test('hash class names', () => {
   ])
 })
 
-test('apply HTML parser options (comment: true)', () => {
```
Comment (sastan) [inline src/shim/server/shim.test.ts:137]: I think we should leave the test to ensure that comments are not removed. And maybe add one for the issue you had (the whitespace thing).

**Review by sastan — CHANGES_REQUESTED** [review]

File: src/shim/server/shim.test.ts (hunk 1 above)
Comment (ggoodman) [reply src/shim/server/shim.test.ts:137]: Added back a html comment test in 962e44f.

I'm actually not sure how to get a minimal repo of the whitespace modifications issue I was seeing.

**Comment by sastan** [comment]
I can confirm the performance improvement:

```
node-html-parser x 6,137 ops/sec ±0.39% (93 runs sampled)
htmlparser2 x 12,614 ops/sec ±0.81% (93 runs sampled)
```

This PR is awesome it doubles the speed and does not modify the HTML except for the class attributes!!!!

Awesome work @ggoodman 🚀

File: src/shim/server/index.ts [hunk 2]
```diff
@@ -55,12 +43,59 @@ export const shim = (markup: string, options: TW | ShimOptions = {}): string =>
   const { tw = defaultTW, ...parserOptions } =
     typeof options === 'function' ? { tw: options } : options
```
Comment (sastan) [inline src/shim/server/index.ts:44]:
```suggestion
  const { tw = defaultTW } = typeof options === 'function' ? { tw: options } : options
```

File: src/shim/server/shim.test.ts (hunk 1 above)
Comment (sastan) [reply src/shim/server/shim.test.ts:137]: I think we are good here.

**Review by sastan — APPROVED** [review]

**Pull request merged**